### ISSUE TITLE
Improve ReadableStream integration with Fetch

### ISSUE THREAD

**Issue by gterzian** [servo/servo#34252]
Follow-up on https://github.com/servo/servo/issues/32898 and https://github.com/servo/servo/issues/29088

After we have re-written our implementation so as not to rely on the deprecated one from Spidermonkey, there is still a lot of work left to improve how streams integrate with other parts of Servo, with Fetch the only example so far. 

1. One part is transmitting the body of a request: https://fetch.spec.whatwg.org/#body-incrementally-read  This is currently done with our Rust code calling directly into a stream via various "native" methods, this should instead happen via a reader on the stream using public methods. See [`TransmitBodyConnectHandler`](https://github.com/servo/servo/blob/495cceb7de813f6f1936d77821e8cf04ca2857cd/components/script/body.rs#L73) 

2. The other part is consuming the body of a response: https://fetch.spec.whatwg.org/#body-fully-read We do this with [`consume_body_with_promise`](https://github.com/servo/servo/blob/495cceb7de813f6f1936d77821e8cf04ca2857cd/components/script/body.rs#L750) again by calling "native" methods on the stream. 

We can start tracking more specific things in here.

**Comment by gterzian** [comment]
cc @Taym95 @wusyong

**Comment by gterzian** [comment]
`/fetch/api/response/response-stream-with-broken-then.any.html` is a particular problem:

In the new setup, we always go through the `Read` dom method of a reader, even when reading a chunk from Rust, and it breaks those tests where `then` is intercepted by JS. 

The current code, using SM APIs, is at https://github.com/servo/servo/blob/9028c8cf67e5c34832c9866871eb3c2bfe25b89d/components/script/dom/readablestream.rs#L262C47-L262C78

It seems to be that `ReadableStreamDefaultReaderRead` passed those tests by using [`PromiseObject::unforgeableResolveWithNonPromise`](https://github.com/servo/mozjs/blob/f68e385deb685e8e74f50e71d878d0889903e378/mozjs-sys/mozjs/js/src/builtin/streams/ReadableStreamReader.cpp#L255), which is at https://github.com/servo/mozjs/blob/f68e385deb685e8e74f50e71d878d0889903e378/mozjs-sys/mozjs/js/src/builtin/Promise.cpp#L4869

I'm not sure if this is currently exposed as an API by https://github.com/servo/mozjs/, but I think the solution would be using that within some sort of `read_native` method that would not go through the public reader API. 

cc @jdm @sagudev

**Comment by sagudev** [comment]
We can check how is done now in firefox, where gecko generally does not access SM internals. I think relevant code is here: https://searchfox.org/mozilla-central/source/dom/fetch/FetchStreamReader.cpp#136

**Comment by jdm** [comment]
This native implementation of Then is the missing piece:
* https://searchfox.org/mozilla-central/source/dom/promise/Promise-inl.h#231

It's called via https://searchfox.org/mozilla-central/source/dom/streams/ReadableStream.cpp#639 in the ReadableStreamCancel implementation.

**Comment by gterzian** [comment]
> This native implementation of Then is the missing piece

I don't think so, because the Gecko code seems to be using a native hander, and that is the same as what we are doing. 

To be clear we're looking at the gecko equivalent of https://github.com/servo/servo/blob/9028c8cf67e5c34832c9866871eb3c2bfe25b89d/components/script/body.rs#L777

So not canceling, but reading from the stream that is the Response body to be consumed. 

To me it seems that in the "native" case, Gecko works directly with `nsIInputStream`, bypassing the JS entirely, see https://searchfox.org/mozilla-central/rev/fb8a1b0beee7dbc1b0a35050ec5c0345f0b09173/dom/fetch/Response.cpp#296

The `FetchStreamReader` seems to be used only to read from a JS stream, and what they call a "DOM generated stream" also goes directly over to the underling input stream see https://searchfox.org/mozilla-central/rev/fb8a1b0beee7dbc1b0a35050ec5c0345f0b09173/dom/fetch/Response.cpp#283

So to sum it up, I think:

- The old SM implementation, when using a "native" underlying source, would rely on this unforgeable promise thing. So we were using that as well, and going through SM even for fully "native" streams. 
- The new Gecko implementation, when working with "native" streams, works directly with the underlying `nsIInputStream`. 
- Our new WIP implementation still uses SM and a promise, but we lost the unforgeable part. 

So I think we have two options:
1. Do something weird with the SM internals.
2. Implement a native async mechanism to resolve promises from native streams.

**Comment by gterzian** [comment]
Some clues here:  https://github.com/web-platform-tests/wpt/commit/4e98c23a16efc0eb6ea4c305e3f48def3cac4643

![Uploading Screenshot 2024-12-04 at 9.30.52 PM.png…]()

**Comment by gterzian** [comment]
I tried setting a null prototype for the returned promise, but the test then failed on a stringification error(I guess because that requires a normal object prototype). Will leave this one alone for now.
